### PR TITLE
Added api.py methods testcases

### PR DESCRIPTION
**Pull request by kdcis**
Added statusBar, getAllInboxMessages, getAllInboxMessageIds, getInboxMessageById, getInboxMessagesByReceiver, trashMessage, trashInboxMessage & addSubscription,  methods testcases

### CODE REVIEW

File: src/testmode_init.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import time
+import uuid
+
+import helper_sql
```
Comment (PeterSurda) [inline src/testmode_init.py:4]: isort

**Review by PeterSurda — CHANGES_REQUESTED** [review]
Fix linter issues

File: src/tests/test_api.py [hunk 2]
```diff
@@ -91,6 +125,31 @@ def test_invalid_method(self):
             'API Error 0020: Invalid method: test'
         )
 
+    def test_statusbar_method(self):
+        """API command 'statusbar'"""
+        self.assertEqual(
```
[the remark this answers is not in the document]
Comment (PeterSurda) [reply src/tests/test_api.py:94]: The API call just puts into `UISignalQueue` and has no API feedback.

File: src/tests/test_api.py [hunk 3]
```diff
@@ -91,6 +89,49 @@ def test_invalid_method(self):
...
+            self.api.statusBar(sample_statusbar_msg),
+            'null'
+        )
```
Comment (PeterSurda) [inline src/tests/test_api.py:97]: wait a bit and then check `uisignal.txt` contents

Comment (PeterSurda) [reply src/tests/test_api.py:97]: try to find how to do it with logging

File: src/tests/test_api.py [hunk 4]
```diff
@@ -15,6 +16,11 @@
 
 from .test_process import TestProcessProto
 
+sample_statusbar_msg = "new status bar message"
+sample_inbox_msg_id = '2c975c515f8b414db5eea60ba57ba455'
+sample_inbox_msg_receiver_address = 'BM-2cWzSnwjJ7yRP3nLEWUV5LisTZyREWSzUK'
+sample_trash_msg_ids = ['27e644765a3e4b2e973ee7ccf958ea20', '51fc5531-3989-4d69-bbb5-68d64b756f5b']
+sample_test_subscription_address = ['BM-2cWQLCBGorT9pUGkYSuGGVr9LzE4mRnQaq', 'BM-GtovgYdgs7qXPkoYaRgrLFuFKz1SFpsw']
```
Comment (g1itch) [inline src/tests/test_api.py:23]: There is already a module `tests.samples` for this: https://github.com/Bitmessage/PyBitmessage/blob/v0.6/src/tests/samples.py

File: src/tests/test_api.py [hunk 5]
```diff
@@ -91,6 +90,49 @@ def test_invalid_method(self):
...
+    def test_statusbar_method(self):
+        """API command 'statusbar'"""
+        self.assertEqual(
+            self.api.statusBar(sample_statusbar_msg),
+            'null'
+        )
+
+    def test_message_inbox(self):
+        """Test message inbox"""
+        # import pdb;pdb.set_trace()
```
Comment (PeterSurda) [inline src/tests/test_api.py:102]: remove pdb

File: src/testmode_init.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+import helper_inbox
+import helper_sql
+
+
+def populate_api_test_data():
+    '''Adding test records in inbox table'''
+    helper_sql.sql_ready.wait()
+
+    test1 = (
```
Comment (PeterSurda) [inline src/testmode_init.py:12]: should refer to variables from `samples.py`

File: src/tests/test_api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (PeterSurda) [reply src/tests/test_api.py:90]: Some other options:
- make a Mock Queue which will log
- make a testmode only API call which will do a non-blocking get from the `UISignalQueues`

File: src/tests/test_api.py [hunk 7]
```diff
@@ -91,6 +85,49 @@ def test_invalid_method(self):
...
+        self.assertEqual(
+            self.api.statusBar(sample_statusbar_msg),
+            'null'
+        )
+
+    def test_message_inbox(self):
+        """Test message inbox"""
+
+        self.assertTrue(
+            len(json.loads(
```
Comment (PeterSurda) [inline src/tests/test_api.py:99]: more accurate check here

File: src/tests/test_api.py [hunk 8]
```diff
@@ -91,6 +85,49 @@ def test_invalid_method(self):
...
+                self.api.getAllInboxMessages())["inboxMessages"]) > 0
+        )
+        self.assertTrue(
+            len(json.loads(
```
Comment (PeterSurda) [inline src/tests/test_api.py:103]: more accurate check here

File: src/tests/test_api.py [hunk 9]
```diff
@@ -91,6 +85,49 @@ def test_invalid_method(self):
...
+                self.api.getAllInboxMessageIds())["inboxMessageIds"]) > 0
+        )
+        self.assertTrue(
+            len(json.loads(
+                self.api.getInboxMessageById(hexlify(sample_inbox_msg_ids[2])))["inboxMessage"]) > 0
```
Comment (PeterSurda) [inline src/tests/test_api.py:108]: also

File: src/tests/test_api.py [hunk 10]
```diff
@@ -91,6 +85,49 @@ def test_invalid_method(self):
...
+        )
+        self.assertTrue(
+            len(json.loads(
```
Comment (PeterSurda) [inline src/tests/test_api.py:111]: also

File: src/tests/test_api.py [hunk 11]
```diff
@@ -193,9 +230,28 @@ def test_addressbook(self):
 
     def test_subscriptions(self):
         """Testing the API commands related to subscriptions"""
+
+        self.assertEqual(
+            self.api.addSubscription(sample_test_subscription_address[0], 'test sub'.encode('base64')),
```
Comment (PeterSurda) [inline src/tests/test_api.py:235]: `'test sub'` should be a variable

**Review by PeterSurda — CHANGES_REQUESTED** [review]
Do the requested changes.

File: src/api.py [hunk 12]
```diff
@@ -1422,11 +1423,69 @@ def HandleAdd(self, a, b):
...
 
+    @command('clearUISignalQueue')
+    def HandleclearUISignalQueue(self):
+        """clear UISignalQueue"""
+        queues.UISignalQueue.queue.clear()
+
     @command('statusBar')
     def HandleStatusBar(self, message):
         """Update GUI statusbar message"""
         queues.UISignalQueue.put(('updateStatusBar', message))
```
Comment (PeterSurda) [inline src/api.py:1434]: just flush here

File: src/api.py [hunk 13]
```diff
@@ -1422,11 +1423,69 @@ def HandleAdd(self, a, b):
...
 
+    @command('getStatusBar')
+    def HandleGetStatusBar(self):
+        """Get GUI statusbar message"""
+        while True:
+            command, data = queues.UISignalQueue.get()
```
Comment (PeterSurda) [inline src/api.py:1440]: non blocking get

File: src/api.py [hunk 14]
```diff
@@ -1422,11 +1423,69 @@ def HandleAdd(self, a, b):
...
+
     @command('statusBar')
     def HandleStatusBar(self, message):
         """Update GUI statusbar message"""
         queues.UISignalQueue.put(('updateStatusBar', message))
 
+    @command('getStatusBar')
+    def HandleGetStatusBar(self):
+        """Get GUI statusbar message"""
+        while True:
```
Comment (PeterSurda) [inline src/api.py:1439]: no while true

**Review by PeterSurda — CHANGES_REQUESTED** [review]

File: src/api.py [hunk 15]
```diff
@@ -1422,11 +1423,22 @@ def HandleAdd(self, a, b):
...
+
     @command('statusBar')
     def HandleStatusBar(self, message):
         """Update GUI statusbar message"""
         queues.UISignalQueue.put(('updateStatusBar', message))
 
+    @command('getStatusBar')
+    def HandleGetStatusBar(self):
+        """Get GUI statusbar message"""
+        command, data = queues.UISignalQueue.get(block=False)
```
Comment (PeterSurda) [inline src/api.py:1439]:
```
_, data = queues.UISignalQueue.get(block=False)
```
Also add exception handler, I think it's `queue.Empty`

**Review by PeterSurda — CHANGES_REQUESTED** [review]

File: src/api.py (hunk 15 above)
Comment (g1itch) [inline src/api.py:1439]: It's a strange decision to add useless API commands only for testing purpose. Also the `UISignalQueue` receives various messages not related to the status bar.

File: src/testmode_init.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
+import time
```
Comment (g1itch) [inline src/testmode_init.py:1]: Why it's here and not in the tests module? End user doesn't need this.

File: src/tests/samples.py [hunk 17]
```diff
@@ -29,3 +29,10 @@
 # Deterministic addresses with stream 1 and versions 3, 4
 sample_deterministic_addr3 = 'BM-2DBPTgeSawWYZceFD69AbDT5q4iUWtj1ZN'
 sample_deterministic_addr4 = 'BM-2cWzSnwjJ7yRP3nLEWUV5LisTZyREWSzUK'
+
+sample_statusbar_msg = "new status bar message"
+sample_inbox_msg_receiver_address = 'BM-2cWzSnwjJ7yRP3nLEWUV5LisTZyREWSzUK'
```
Comment (g1itch) [inline src/tests/samples.py:34]: Why double the sample_deterministic_addr4 3 lines above?

File: src/tests/samples.py [hunk 18]
```diff
@@ -29,3 +29,10 @@
...
+sample_inbox_msg_ids = ['27e644765a3e4b2e973ee7ccf958ea20', '51fc5531-3989-4d69-bbb5-68d64b756f5b',
+                        '2c975c515f8b414db5eea60ba57ba455', 'bc1f2d8a-681c-4cc0-9a12-6067c7e1ac24']
+sample_test_subscription_address = ['BM-2cWQLCBGorT9pUGkYSuGGVr9LzE4mRnQaq', 'BM-GtovgYdgs7qXPkoYaRgrLFuFKz1SFpsw']
```
Comment (g1itch) [inline src/tests/samples.py:37]: Why do you need the second address?

File: src/tests/test_api.py [hunk 19]
```diff
@@ -91,6 +86,53 @@ def test_invalid_method(self):
...
+            len(json.loads(
+                self.api.getInboxMessageById(hexlify(sample_inbox_msg_ids[2])))["inboxMessage"]) == 1
+        )
+        self.assertTrue(
+            len(json.loads(
+                self.api.getInboxMessagesByReceiver(sample_inbox_msg_receiver_address))["inboxMessages"]) == 2
+        )
+
+    def test_delete_message(self):
+        """Issuing nonexistent command 'test'"""
```
Comment (g1itch) [inline src/tests/test_api.py:121]: ??

File: src/tests/test_api.py [hunk 20]
```diff
@@ -91,6 +86,53 @@ def test_invalid_method(self):
             'API Error 0020: Invalid method: test'
         )
 
+    def test_statusbar_method(self):
+        """API command 'statusbar'"""
```
Comment (g1itch) [inline src/tests/test_api.py:90]: wut?

**Review by g1itch — COMMENTED** [review]
Another portion of garbage );

File: src/api.py (hunk 15 above)
Comment (PeterSurda) [reply src/api.py:1439]: Ideally this command would only be available in test mode, as well as the flush one. But better to have some testing. There were several bugs related to queues in the past.

File: src/testmode_init.py (hunk 16 above)
Comment (PeterSurda) [reply src/testmode_init.py:1]: this file will be refactored once #1794 is merged. Currently there is no nice way to put it inside the tests.

File: src/tests/test_process.py [hunk 21]
```diff
@@ -61,7 +61,6 @@ def setUpClass(cls):
         starttime = int(time.time()) - 0.5
         cls.process = psutil.Popen(
             cls._process_cmd, stderr=subprocess.STDOUT)  # nosec
-
```
Comment (PeterSurda) [inline src/tests/test_process.py:64]: revert

File: src/tests/samples.py (hunk 18 above)
Comment (PeterSurda) [reply src/tests/samples.py:43]: That's the "announcements" broadcast.

File: src/api.py [hunk 22]
```diff
@@ -1422,11 +1424,25 @@ def HandleAdd(self, a, b):
         """Test two numeric params"""
         return a + b
 
+    @command('clearUISignalQueue')
```
Comment (PeterSurda) [inline src/api.py:1427]: something to make this only available in test mode

File: src/api.py [hunk 23]
```diff
@@ -1422,11 +1424,25 @@ def HandleAdd(self, a, b):
...
+    def HandleclearUISignalQueue(self):
+        """clear UISignalQueue"""
+        queues.UISignalQueue.queue.clear()
+
     @command('statusBar')
     def HandleStatusBar(self, message):
         """Update GUI statusbar message"""
         queues.UISignalQueue.put(('updateStatusBar', message))
 
+    @command('getStatusBar')
```
Comment (PeterSurda) [inline src/api.py:1437]: also only in test mode

File: src/api.py [hunk 24]
```diff
@@ -308,6 +310,9 @@ def __init__(self, *aliases):
         self.aliases = aliases
 
     def __call__(self, func):
+        if func.__name__ in ["HandleclearUISignalQueue", "HandleGetStatusBar"] \
```
Comment (PeterSurda) [inline src/api.py:313]: create a constant, say
```
TESTMODE_FUNCTIONS = ["HandleclearUISignalQueue", "HandleGetStatusBar"]

if func.__name__ in TESTMODE_FUNCTIONS and not state.testmode:
    return None
```

File: src/api.py (hunk 22 above)
Comment (PeterSurda) [reply src/api.py:1430]: My preferred way would be to create an additional decorator for testmode functions.

File: src/tests/samples.py [hunk 25]
```diff
@@ -35,3 +35,9 @@
 sample_deterministic_addr4 = 'BM-2cWzSnwjJ7yRP3nLEWUV5LisTZyREWSzUK'
 sample_daddr3_512 = 18875720106589866286514488037355423395410802084648916523381
 sample_daddr4_512 = 25152821841976547050350277460563089811513157529113201589004
+
+sample_statusbar_msg = "new status bar message"
+sample_inbox_msg_ids = ['27e644765a3e4b2e973ee7ccf958ea20', '51fc5531-3989-4d69-bbb5-68d64b756f5b',
+                        '2c975c515f8b414db5eea60ba57ba455', 'bc1f2d8a-681c-4cc0-9a12-6067c7e1ac24']
+sample_test_subscription_address = ['BM-2cWQLCBGorT9pUGkYSuGGVr9LzE4mRnQaq', 'BM-GtovgYdgs7qXPkoYaRgrLFuFKz1SFpsw']
```
Comment (PeterSurda) [inline src/tests/samples.py:42]: maybe say in a comment that the second address is the announcement broadcast?

**Review by PeterSurda — CHANGES_REQUESTED** [review]
Just a minor change.

File: src/tests/test_api.py [hunk 26]
```diff
@@ -91,6 +85,53 @@ def test_invalid_method(self):
...
+        """Test message inbox methods"""
+        self.assertTrue(
+            len(json.loads(
+                self.api.getAllInboxMessages())["inboxMessages"]) == 2
+        )
+        self.assertTrue(
+            len(json.loads(
+                self.api.getAllInboxMessageIds())["inboxMessageIds"]) == 2
+        )
+        self.assertTrue(
```
Comment (PeterSurda) [inline src/tests/test_api.py:110]: assertEqual

File: src/tests/test_api.py [hunk 27]
```diff
@@ -91,6 +85,53 @@ def test_invalid_method(self):
...
+            len(json.loads(
+                self.api.getInboxMessageById(hexlify(sample_inbox_msg_ids[2])))["inboxMessage"]) == 1
+        )
+        self.assertTrue(
```
Comment (PeterSurda) [inline src/tests/test_api.py:114]: assertEqual

File: src/tests/test_api.py [hunk 28]
```diff
@@ -91,6 +85,58 @@ def test_invalid_method(self):
...
+        )
+        self.assertEqual(
+            self.api.trashInboxMessage(hexlify(sample_inbox_msg_ids[1])),
+            'Trashed inbox message (assuming message existed).'
+        )
+        self.assertEqual(
+            len(json.loads(self.api.getAllInboxMessageIds())["inboxMessageIds"]),
+            messages_before_delete - 2
+        )
+
```
Comment (PeterSurda) [inline src/tests/test_api.py:139]: undelete the two messages (make a new test-only API call for this)

Comment (PeterSurda) [reply src/tests/test_api.py:139]: something like

```
# tearDown
self.assertTrue(json.loads(self.api.undeleteMessage(sample_inbox_msg_ids[0])["result"])
self.assertTrue(json.loads(self.api.undeleteMessage(sample_inbox_msg_ids[1])["result"])
```

File: src/api.py [hunk 29]
```diff
@@ -302,12 +304,27 @@ def __new__(mcs, name, bases, namespace):
...
+        return command(self.aliases[0]).__call__(func)
+
+
 class command(object):  # pylint: disable=too-few-public-methods
     """Decorator for API command method"""
     def __init__(self, *aliases):
         self.aliases = aliases
 
     def __call__(self, func):
+
```
Comment (PeterSurda) [inline src/api.py:327]: why space?

File: src/tests/test_api.py [hunk 30]
```diff
@@ -91,6 +85,71 @@ def test_invalid_method(self):
...
+            sample_statusbar_msg
+        )
+
+    def test_message_inbox(self):
+        """Test message inbox methods"""
+        self.assertEqual(
+            len(json.loads(
+                self.api.getAllInboxMessages())["inboxMessages"]),
+            4,
+            json.loads(self.api.getAllInboxMessages())["inboxMessages"]
```
Comment (PeterSurda) [inline src/tests/test_api.py:106]: Maybe add a comment explaining this is the custom error message to aid in debugging.

**Review by PeterSurda — APPROVED** [review]

**Pull request merged**